worker-workflow/src/main/java/com/github/cafdataprocessing/TransformedWorkflowCache.java

```diff
@@ -0,0 +1,173 @@
+        this.dataStore = dataStore;
+        this.processingApiUrl = processingApiUrl;
+        workflowCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(workflowCachePeriod.get(ChronoUnit.SECONDS), TimeUnit.SECONDS)
+                .build(new CacheLoader<TransformedWorkflowCacheKey, TransformWorkflowResult>() {
+                    @Override
+                    public TransformWorkflowResult load(TransformedWorkflowCacheKey key)
+                            throws ApiException, DataStoreException, WorkflowTransformerException {
+                        return transformWorkflow(key);
+                    }
+                });
+    }
+
+    /**
+     * Retrieves the result of a workflow transformation using combination of the provided workflow ID, project ID and
+     * partial storage reference as a key. If no entry is found for the key then an attempt will be made to transform the
+     * workflow with that result being returned.
+     * @param workflowId ID of workflow indicating the transformed workflow result to retrieve.
+     * @param projectId project ID indicating the transformed workflow result to retrieve.
+     * @param outputPartialReference partial reference used in storing transformed workflow.
+     * @return returns the transformed workflow result associated with the workflow ID, project ID & partial reference.
+     * @throws ApiException if a failure occurs communicating with processing API during load of transformed workflow result.
+     * @throws DataStoreException if a failure occurs storing a transformed workflow during load.
+     * @throws ExecutionException if a checked exception occurs during load of transformed workflow result that is not
+     * expected by this method.
+     * @throws UncheckedExecutionException if an unchecked exception occurs during load of transformed workflow result
+     * that is not expected by this method.
+     * @throws WorkflowTransformerException if a failure occurs during transformation of workflow during load.
+     */
+    public TransformWorkflowResult getTransformWorkflowResult(final long workflowId, final String projectId,
+                                                              final String outputPartialReference)
+            throws ApiException, ExecutionException, DataStoreException, UncheckedExecutionException, WorkflowTransformerException {
+        final TransformedWorkflowCacheKey cacheKey = buildCacheKey(workflowId, projectId, outputPartialReference);
+        try {
+            return workflowCache.get(cacheKey);
+        }
+        catch(final ExecutionException e) {
+            // throw specific exception types that caused ExecutionException if they are known to be thrown during cache
+            // load
+            final Throwable cause = e.getCause();
+            if(cause instanceof ApiException ) {
+                LOG.error("Processing API exception when trying to retrieve workflow.", cause);
```

Comment:
Are we double-logging here?  I see that you already logged this on the other side in the `transformWorkflow` function.  What is the correct level for it to be logged at?

Reply:
Yeah that's double-logging. Removed this and left it in the transform workflow as error level.